File: generateProjectedPoints.cpp
// LibQPEP: A Library for Globally Optimal Solving Quadratic Pose Estimation Problems,
//          It also gives highly accurate uncertainty description of the solutions.
//
// Authors: Jin Wu and Ming Liu
// Affiliation: Hong Kong University of Science and Technology (HKUST)
// Emails: jin_wu_uestc@hotmail.com; eelium@ust.hk
// Reference: Wu, J., et al. (2022) Quadratic Pose Estimation Problems: 
//                                  Globally Optimal Solutions, 
//                                  Solvability/Observability Analysis,
//                                  and Uncertainty Description.
//                                  IEEE Transactions on Robotics.
//                                  https://doi.org/10.1109/TRO.2022.3155880
//
//
// generateProjectedPoints.cpp: Generate image points for PnP problems


#include "generateProjectedPoints.h"

void generateProjectedPoints(std::vector<Eigen::Vector2d>& image_pt,
                             std::vector<double>& s,
                             const std::vector<Eigen::Vector3d>& world_pt,
                             const Eigen::Matrix3d& K,
                             const Eigen::Matrix3d& R,
                             const Eigen::Vector3d& t)
{
    int numPoints = world_pt.size();

    for(int i = 0; i < numPoints; ++i)
    {
        s.push_back(0);
    }

    for(int i = 0; i < numPoints; ++i)
    {
        Eigen::Vector4d world_point(world_pt[i](0), world_pt[i](1), world_pt[i](2), 1);
        Eigen::Matrix<double, 3, 4> cameraMatrix;
        cameraMatrix << R.transpose(), t;
        cameraMatrix = K * cameraMatrix;
        Eigen::Vector3d projectedPoint = cameraMatrix * world_point;
        s[i] = projectedPoint(2);
        Eigen::Vector2d projectedPoint_(projectedPoint(0), projectedPoint(1));
        image_pt.push_back(projectedPoint_ / s[i]);
    }
}

File: MATLAB/utils/generateProjectedPoints_.m
% 
% LibQPEP: A Library for Globally Optimal Solving Quadratic Pose Estimation Problems (QPEPs),
%          It also gives highly accurate uncertainty description of the solutions.
%
%
% Article: 
%      Wu, J., Zheng, Y., Gao, Z., Jiang, Y., Hu, X., Zhu, Y., Jiao, J., Liu, M. (2020)
%           Quadratic Pose Estimation Problems: Unified Solutions, 
%           Solvability/Observability Analysis and Uncertainty Description 
%           in A Globally Optimal Framework.
%
%
% Authors:      Jin Wu and Ming Liu
% Affiliation:  Hong Kong University of Science and Technology (HKUST)
% Emails:       jin_wu_uestc@hotmail.com; eelium@ust.hk
% Websites:     https://zarathustr.github.io
%               https://ram-lab.com



function [ProjectedPoints, s] = generateProjectedPoints_(world_pt, K, R, t)
numPoints = size(world_pt, 1);
ProjectedPoints = zeros(numPoints, 2);
s = zeros(numPoints);
for i = 1 : numPoints
    world_point = [world_pt(i, :), 1]; % homogeneous coordinates

    cameraMatrix = [R; t'] * K;
    projectedPoint = world_point * cameraMatrix;
    s(i) = projectedPoint(3);
    projectedPoint = projectedPoint(1 : 2) ./ s(i);
    ProjectedPoints(i, :) = projectedPoint;
end
end